### PR TITLE
Scoring

### PR DESCRIPTION
**Pull request by raoulmillais**
This implements the functionality decribed in #173 ... the attack container has been pushed with the tag `scoring-mvp`.

**To test**

* Clone and checkout the scoring branch
* Recreate your infra with `--attack-container-tag=scoring-mvp`
* Check each of the logic branches described in the ticket #173 

Try starting tasks showing hints starting new tasks, going back and forth etc and make sure the messaging feels right and there are no errors etc :)

Other than obvious bugs, Im particularly interested in feedback on the user experience and messaging.  Unleash your creative destruction :)

This resolves both #173  and partially #172 (this implementation does not ask the user to confirm they want to incur the penalty to see the hint)

### Update

I have done a large refactoring to make the code clearer and to add a boatload of automated tests to check all the edge cases I could think of.  Unfortunately this makes this a bit of a monster pull request, but the behaviour is the same and the cases I have covered can be seen by running the tests in verbose mode `cd tools/scenario-tools/ && npm test -- --verbose`:

```
<snip>
  ✔ tasks_test › getCurrentTask returns the current task
  ✔ tasks_test › getCurrentTask returns undefined when no current task
  ✔ tasks_test › updateProgressWithNewTask sets current_task
  ✔ tasks_test › updateProgressWithNewTask initialises task progress
  ✔ tasks_test › updateProgressWithNewTask does not overwrite existing progress
  ✔ tasks_test › processResponse throws when answer is not defined
  ✔ tasks_test › processResponse returns false when answer is cancel
  ✔ tasks_test › processResponse sets score to skip when answer is no
  ✔ tasks_test › processResponse sets score when answer is yes
  ✔ tasks_test › processTask warns for invalid task and returns false
  ✔ tasks_test › endTask warns if the user has not started a task
  ✔ tasks_test › startTask warns if user tries to start the same task
  ✔ tasks_test › startTask initialises progress when user has no progress
  ✔ tasks_test › startTask updates current_task and doesnt rescore if already scored
  ✔ tasks_test › processTask writes current_task to progress
  ✔ tasks_test › endTask returns false if the user cancels
  ✔ tasks_test › endTask skips scoring and ends task when user says no to scoring
  ✔ tasks_test › endTask sets score and ends task when user says yes to scoring
  ✔ tasks_test › startTask updates current_task and skips scoring if answer is no
  ✔ tasks_test › startTask updates current_task and scores if answer is yes
<snip>
```

### CODE REVIEW

File: attack/scripts/bashrc [hunk 1]
```diff
@@ -249,6 +249,14 @@ welcome() {
   info "  \$ next_hint"
   info "  \$ show_hints"
   info ' '
+  info 'To end a task, you can start a new task or:'
```
Comment (pi-unnerup) [inline attack/scripts/bashrc]: welcome message should maybe mention that you can only finish a task once. 
also should you be able to do task 1 after task 2? In all cases task two will give the solution to task 1 and in most cases preclude it by preventing it

File: attack/scripts/bashrc [hunk 2]
```diff
@@ -249,6 +249,14 @@ welcome() {
...
+  info ' '
+  info "  \$ end_task"
```
Comment (pi-unnerup) [inline attack/scripts/bashrc]: nice

File: tools/scenario-tools/lib/hints.js [hunk 3]
```diff
@@ -58,15 +59,21 @@ function nextHint (task, taskspath = TASKS_FILE_PATH,
...
-    hintIndex = progress[task] = 0
+    progress[task] = {}
+  }
+
+  if (progress[task].lastHintIndex === undefined) {
+    hintIndex = progress[task].lastHintIndex = 0
   } else {
-    const { tasks } = loadYamlFile(taskspath)
+    hintIndex = progress[task].lastHintIndex = progress[task].lastHintIndex + 1
+  }
```
Comment (pi-unnerup) [inline tools/scenario-tools/lib/hints.js]: you should be able to show_hints even after all hints have been shown

**Comment by abdullahgarcia** [comment]
I've checked the PR, functionally looks fine.

I would only modify the message regarding scoring in the bashrc file:

"Remember, when you start a task you will only have one
opportunity to be scored on that task. Once it is ended or if you switch to another task,
you will either be scored or marked as 'skipping scoring'. You can still
return to the task but scoring will be fixed to the result of your first
attempt.

To start the first task:"
